File: node-local-distributed-cache-api/src/cep/client/ViaCepClient.js
import axios from "axios";

import { VIA_CEP_URI } from "../../config/secrets.js";

class ViaCepClient {
  async findByCep(cep) {
    let response = {};
    console.info(`Buscando dados da API do ViaCep para o CEP ${cep}.`);
    await axios
      .get(VIA_CEP_URI(cep))
      .then((res) => {
        let apiResponse = res.data;
        response = {
          cep: apiResponse.cep,
          logradouro: apiResponse.logradouro,
          complemento: apiResponse.complemento,
          bairro: apiResponse.bairro,
          localidade: apiResponse.localidade,
          uf: apiResponse.uf,
        };
        console.log(
          `Retornando dados da API do ViaCep para o CEP ${cep}: ${JSON.stringify(
            apiResponse
          )}`
        );
      })
      .catch((err) => {
        console.error(`Erro ao chamar a API do ViaCep para o CEP ${cep}`);
        console.error(err);
      });
    return response;
  }
}

export default new ViaCepClient();

File: java-local-distributed-cache-api/src/main/java/br/com/cache/localdistributedcache/cep/client/ViaCepClient.java
package br.com.cache.localdistributedcache.cep.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Slf4j
@Component
public class ViaCepClient {

    @Autowired
    private WebClient webClient;

    @Autowired
    private ObjectMapper objectMapper;

    @Value("${app-config.client.viacep}")
    private String viaCepUri;

    public Mono<String> findByCep(String cep) {
        log.info("Buscando dados da API do ViaCep para o CEP {}", cep);
        return webClient
            .get()
            .uri(buildUri(cep))
            .retrieve()
            .bodyToMono(String.class)
            .onErrorResume(throwable -> {
                log.error("Erro ao chamar a API do ViaCep para o CEP {}", cep, throwable);
                return Mono.empty();
            })
            .doOnNext(response -> log.info("Retornando dados da API do ViaCep para o CEP {}: {}", cep, response));
    }

    public String buildUri(String cep) {
        return String.format(viaCepUri, cep);
    }
}
